keras_nlp/layers/sine_positional_embedding.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Sinusoidal position embedding layer."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class SinePositionEncoding(keras.layers.Layer):
+    """Sinusoidal positional encoding layer.
+
+    This layer calculates the position encoding as a mix of sine and cosine
+    functions with geometrically increasing wavelengths. Defined and formulized
+    in "Attention is All You Need", section 3.5.
+    (https://arxiv.org/abs/1706.03762).
+
+    Args:
+        min_frequency: The minimum frequency for each position embedding,
+                       initalized to 1.0e-4.
+
+    Usage:
+        Pass as input an embedded token tensor, where the last dimension is
+        assumed to be the feature dim, and the second to last dimension is
+        assumed to be the sequence dimension. This layer will return a
+        positional encoding the same size as the embedded token tensor, which
+        can be added directly to the embedded token tensor.
+
+    Example:
+    ```python
+    # create a simple embedding layer with sinusoidal positional encoding
+    inputs = keras.Input((100, ), dtype=tf.float32)
+    embedding = keras.layers.Embedding(input_dim=1000, output_dim=32)(inputs)
+    positional_encoding = keras_nlp.layers.SinePositionEncoding()(embedding)
+    outputs = keras.layers.Add()([embedding, positional_encoding])
+    ```
+
+    References:
+      [Attention is All You Need](https://arxiv.org/abs/1706.03762)
+
+    """
+
+    def __init__(self, min_frequency: float = 1.0e-4, **kwargs):
+
+        super().__init__(**kwargs)
+        self.min_frequency = min_frequency
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "min_frequency": self.min_frequency,
+            }
+        )
+        return config
+
+    def call(self, inputs):
+        """Forward pass of the SinePositionEncoding.
+
+        Args:
+            inputs: a Tensor.
+
+        Returns:
+            A Tensor of the same shape as the `inputs` containing the sinusoidal
+            positional encoding weights for the input sequence.
+        """
+        input_shape = tf.shape(inputs)
+        # length of sequence is the second last dimension of the inputs
+        seq_length = input_shape[-2]
+        hidden_size = input_shape[-1]
+        position = tf.cast(tf.range(seq_length), self.compute_dtype)
```

Comment:
Maybe I am making mistakes here, but I don't think these `self.compute_dtype` casting is necessary? Every casting is making copies, so we should only call it when required.

Reply:
I checked and couldn't find any casting that could be removed.

Reply:
Oh tf.range takes a dtype argument. Maybe we should do that everywhere instead of this cast(range) pattern. That would remove almost all of them.

Reply:
I see, thanks for the explanation! Also seems tf.range(10, dtype=tf.float16) fails, so let's just use `tf.cast()`.